[SPARK-35423][ML] PCA results should be consistent, If the Matrix contains both Sparse and Dense vectors

shahidki31 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error message, please read the guideline first:
     https://spark.apache.org/error-message-guidelines.html
-->

### What changes were proposed in this pull request?
If the dataset contains mix of sparse and dense vectors output of PCA seems different. The issue here is we check only the first row's Vector type. If the first row is dense and rest all the row's are sparse, we compute PCA based on dense path. Similarly, if only first row in Sparse and rest all the rows are dense, we compute based on Sparse computation path.

Following datasets will produce different results with PCA, even though the data is same, except first row type is sparse.
```
val data1 = Array(
  Vectors.sparse(5, Seq((1, 1.0), (3, 7.0))),
  Vectors.dense(2.0, 0.0, 3.0, 4.0, 5.0),
  Vectors.dense(4.0, 0.0, 0.0, 6.0, 7.0)
)
```

```
+-----------------------------------------------------------+
|pcaFeatures                                                |
+-----------------------------------------------------------+
|[1.6485728230883807,-4.013282700516296,-5.524543751369388] |
|[-4.645104331781534,-1.1167972663619026,-5.524543751369387]|
|[-6.428880535676489,-5.337951427775355,-5.524543751369389] |
+-----------------------------------------------------------+

```
```
val data1 = Array(
  Vectors.dense(0.0, 1.0, 0.0, 7.0, 0.0 ),
  Vectors.dense(2.0, 0.0, 3.0, 4.0, 5.0),
  Vectors.dense(4.0, 0.0, 0.0, 6.0, 7.0)
)
```

```
+------------------------------------------------------------+
|pcaFeatures                                                 |
+------------------------------------------------------------+
|[1.6485728230883814,-4.0132827005162985,-1.0091435193998504]|
|[-4.645104331781533,-1.1167972663619048,-1.0091435193998501]|
|[-6.428880535676488,-5.337951427775359,-1.009143519399851]  |
+------------------------------------------------------------+
```

<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
To fix inconsistent result if dataset contains both sparse and dense vectors. We need to treat the entire metrics as Sparse ONLY if all the rows are sparse. Otherwise we need to consider the matrix as dense. This PR can be a followup for the PR: https://github.com/apache/spark/pull/23126


### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Added UTs

shahidki31 committed:
PCA results should be consistent, If the Matrix contains both Sparse and Dense vectors

shahidki31 commented:
cc @srowen Kindly review.

SparkQA commented:
**[Test build #139179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139179/testReport)** for PR 32734 at commit [`bdaed96`](https://github.com/apache/spark/commit/bdaed96752392bc25999a6ebb4567922c0dffe1b).

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala:
```diff
@@ -439,7 +439,8 @@ class RowMatrix @Since("1.0.0") (
 
-    if (rows.first().isInstanceOf[DenseVector]) {
+    // If all the rows are sparse vectors, then compute based on `computeSparseVectorCovariance`.
+    if (!rows.filter(_.isInstanceOf[DenseVector]).isEmpty()) {
```
How about `rows.exists(_.isInstanceOf[DenseVector])`? might short-circuit faster. This could be expensive I guess if it really is all sparse, but correctness is more important.

shahidki31 replied:
I think `rows` is an RDD of Vector. There is no `exists` defined in the RDD class. Please correct me if I am wrong.

Also in the above step, we do expensive operations will computing summary, so not sure this filter is expense comparatively.

srowen replied:
Ah oops, for some reason thinking this was a Seq.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43698/

SparkQA commented:
**[Test build #139179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139179/testReport)** for PR 32734 at commit [`bdaed96`](https://github.com/apache/spark/commit/bdaed96752392bc25999a6ebb4567922c0dffe1b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139179/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43698/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43698/

srowen reviewed (commented):
Looks reasonable to me. @zhengruifeng WDYT?

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala:
```diff
@@ -439,7 +439,8 @@ class RowMatrix @Since("1.0.0") (
 
-    if (rows.first().isInstanceOf[DenseVector]) {
+    // If all the rows are sparse vectors, then compute based on `computeSparseVectorCovariance`.
+    if (!rows.filter(_.isInstanceOf[DenseVector]).isEmpty()) {
```
If all the rows are sparse vectors, does `rows.filter(_.isInstanceOf[DenseVector]).isEmpty()` take a whole pass on the input rdd?

shahidki31 replied:
@zhengruifeng Yes.

zhengruifeng commented:
current estimation of data sparsity/density by `first().isInstanceOf[DenseVector]` is too simple

`!rows.filter(_.isInstanceOf[DenseVector]` is better, but it still make following two datasets generate differenct results:

```
val data1 = Array(
  Vectors.dense(0.0, 1.0, 0.0, 7.0, 0.0 ),
  Vectors.dense(2.0, 0.0, 3.0, 4.0, 5.0),
  Vectors.dense(4.0, 0.0, 0.0, 6.0, 7.0)
)

val data2 = data1.map(_.toSparse)
```

if we want to do it more precisely, what about: 
1, add a meta data `sparsity` or `nnz` in vector column, like existing `numFeatures`;
2, impls can fetch this meta if any;
3, otherwise, compute average nnz, and use it to detect sparsity;

zhengruifeng commented:
Besides PCA, there are other impls that rely on the estimation of sparsity. what about just adding a boolean param named `preferSpare` to choose the code path?

shahidki31 commented:
Thanks @zhengruifeng for the review. I will explore more in that area.

>1, add a meta data sparsity or nnz in vector column, like existing numFeatures;

So you mean if the number of zeros in more then consider that vector column as sparse, and If all the vector columns are sparse, then proceed with the sparse based implementation. right?

zhengruifeng commented:
> So you mean if the number of zeros is more in a vector column, then consider that vector column as sparse, and If all the vector columns are sparse, then proceed with the sparse based implementation. right?

This is just an idea. I mean compute average nnz, if nnz_avg < numFeatues * factor(0.5?), the proceduce perfer the code path prefer sparse dataset, otherwise another path.

shahidki31 committed:
Address comment

shahidki31 commented:
Hi @zhengruifeng @srowen  , I have updated following
1. Since Summary metrics already have nnz metrics, we can leverage that to find sparsity. Already the method is computing summary for columns.
2. Updated the UTs

Please review.

SparkQA commented:
**[Test build #139294 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139294/testReport)** for PR 32734 at commit [`1280116`](https://github.com/apache/spark/commit/12801163b1c5bb64cb4dd8491f844b7e7a1b99d7).

shahidki31 committed:
Address comment

SparkQA commented:
**[Test build #139295 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139295/testReport)** for PR 32734 at commit [`55e83ec`](https://github.com/apache/spark/commit/55e83ec1058034539abe295d9ff97d8aed4fc971).

shahidki31 commented:
Hi, @zhengruifeng @srowen 
I have updated the PR based on the following
1. Added a method for checking sparsity of the vector in Vectors.scala class. 
2. In RowMatrix class, added a method to check if all the rows are sparse.
3. Updated UTs
Please review

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43818/

SparkQA commented:
**[Test build #139297 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139297/testReport)** for PR 32734 at commit [`ee82258`](https://github.com/apache/spark/commit/ee82258e67125156aad6333ac174fae678dc781b).

SparkQA commented:
**[Test build #139294 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139294/testReport)** for PR 32734 at commit [`1280116`](https://github.com/apache/spark/commit/12801163b1c5bb64cb4dd8491f844b7e7a1b99d7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139294/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43818/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43818/

SparkQA commented:
**[Test build #139295 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139295/testReport)** for PR 32734 at commit [`55e83ec`](https://github.com/apache/spark/commit/55e83ec1058034539abe295d9ff97d8aed4fc971).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43820/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43820/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139295/

SparkQA commented:
**[Test build #139297 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139297/testReport)** for PR 32734 at commit [`ee82258`](https://github.com/apache/spark/commit/ee82258e67125156aad6333ac174fae678dc781b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139297/

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala:
```diff
@@ -421,6 +421,11 @@ class RowMatrix @Since("1.0.0") (
 
+  // The matrix is sparse, if all the rows has sparsity more than 0.5.
+  private def isSparseMatrix: Boolean = {
+    rows.filter(row => row.sparsity() < 0.5).isEmpty()
```
I think this could be fine but why 0.5? do we not still have similar problems for less-sparse sparse input? I feel like this isn't fixing the underlying problem - does anyone know why PCA is giving different results for sparse input?

shahidki31 replied:
Hi @srowen , The underlying assumptions behind the PR are following,
1. For Sparse data, both the methods gives accurate results, but for dense data `computeDenseVectorCovariance` gives more accurate results.
2. `computeDenseVectorCovariance` is costlier than `computeSparseVectorCovariance`.

So based on the above points, we need to restrict the computeSparseVectorCovariance only to sparse matrix.
But how to find the sparse matrix is the problem. In this PR, if more than 50% of the matrix/vector values are sparse, then consider that matrix as sparse.

The 0.5 factor can change to 0.6 or 0.75 to increase the sparsity.

srowen replied:
Do you have insight into why the dense computation is more accurate? it shouldn't be in theory. I am OK with the change if it's a band-aid that improves results, but not quite sure why this is happening.

shahidki31 replied:
@KyleLi1985 has done experiments with Sparse and dense data https://github.com/apache/spark/pull/23126#issuecomment-441658381 in the PR: https://github.com/apache/spark/pull/23126. 

The conclusion was for dense data `computeSparseVectorCovariance` doesn't give good accuracy. That is why we have moved to `computeDenseVectorCovariance`. In this PR, I am trying to classify which one to choose dense and sparse path. 
 I can do more experimentation on sparse and dense datasets to verify the accuracy part.

shahidki31 replied:
>Do you have insight into why the dense computation is more accurate? it shouldn't be in theory.

Yes, For dense case we are using the actual formula to compute the covariance (which is costly). For sparse case we do approximation.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala:
```diff
@@ -438,8 +443,8 @@ class RowMatrix @Since("1.0.0") (
-    if (rows.first().isInstanceOf[DenseVector]) {
+    // If all the rows are sparse vectors, then compute based on `computeSparseVectorCovariance`.
+    if (!isSparseMatrix) {
       computeDenseVectorCovariance(mean, n, m)
```
OK, I was going to say, why do we treat these two cases differently? but the code already does. I think this is a better test, so I support the change.

shahidki31 replied:
So you mean, we should use `computeDenseVectorCovariance` for both sparse and dense data and remove the `computeSparseVectorCovariance` method?

srowen replied:
No I think the change is reasonable. Just wondering why it ever treated these two paths so differently. But that is not the issue here.

zhengruifeng replied:
I will take a look at the two different code pathes.
As to this PR, what about just adding a param (values are "dense"/"sparse"/"auto"(just keep curent behavior)) to control which impl is used?

shahidki31 replied:
Sure, I will update

zhengruifeng replied:
@shahidki31   This is just a idea. What is your and @srowen  thoughts on it?

There are several similar existing places in ML, like `ANOVATest`.
And I am also thinking about whether to make two impls (for dense cases using BLAS, for sparse cases using triangle-inequality) in KMeans.

srowen replied:
Hm, maybe, but I don't know whether the caller can meaningfully figure out how to set it. I think this change is OK - but how did you determine the correct answer for this case? In the UT?

shahidki31 replied:
@srowen I have done experimentation on 10x10 dummy data with 1's and 0's. I have computed the Eigen values and Eigen vectors using both the methods. Following are the observation

1. For both sparse and dense, strong Eigen value difference is almost negligible (diff is factor of 10^-16). Huge variations come in the weak Eigen values and corresponding Eigen vectors.
2. Users are mostly interested in strong Eigen values. So, fluctuations in the weak Eigen values doesn't affect end users much.

srowen commented:
Jenkins test this please

SparkQA commented:
**[Test build #139503 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139503/testReport)** for PR 32734 at commit [`ee82258`](https://github.com/apache/spark/commit/ee82258e67125156aad6333ac174fae678dc781b).

SparkQA commented:
**[Test build #139503 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139503/testReport)** for PR 32734 at commit [`ee82258`](https://github.com/apache/spark/commit/ee82258e67125156aad6333ac174fae678dc781b).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139503/

shahidki31 commented:
Retest this please

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44028/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44028/

SparkQA commented:
**[Test build #139505 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139505/testReport)** for PR 32734 at commit [`ee82258`](https://github.com/apache/spark/commit/ee82258e67125156aad6333ac174fae678dc781b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44030/

SparkQA commented:
**[Test build #139505 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139505/testReport)** for PR 32734 at commit [`ee82258`](https://github.com/apache/spark/commit/ee82258e67125156aad6333ac174fae678dc781b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139505/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44030/

srowen commented:
Merged to master
